### PR TITLE
billing: automated billing & pay as you go usage billing

### PR DESCRIPTION
**Pull request by gioelecerati**
**What does this pull request do? Explain your changes. (required)**

- Create new products
- Create pay as you go products
- Report overusage
- Display upcoming invoices
- Automated billing

**Specific updates (required)**

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Screenshots (optional)**

**Checklist**

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/api/src/controllers/stripe.ts [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+  // Return a response to acknowledge receipt of the event
+  return res.sendStatus(200);
+});
+
+async function sleep(millis) {
+  return new Promise((resolve) => setTimeout(resolve, millis));
+}
+
+// Migrate existing users to stripe
+app.post("/migrate-users-to-stripe", async (req, res) => {
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Let's delete this endpoint

File: packages/api/src/controllers/stripe.ts [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+interface OverUsageItem {
+  units: number;
+  total: number;
+}
+
+interface BillingUsageData {
+  DeliveryUsageMins: number;
+  TotalUsageMins: number;
+  StorageUsageMins: number;
+}
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: can remove

File: packages/api/src/controllers/stripe.ts [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+
+interface OverUsageItem {
+  units: number;
+  total: number;
+}
+
+interface BillingUsageData {
+  DeliveryUsageMins: number;
+  TotalUsageMins: number;
+  StorageUsageMins: number;
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Reminder to make it lowerCamel

File: packages/api/src/controllers/stripe.ts [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+  scale_1: "prod_O9XtcfOSMjSD5L",
+};
+
+export const reportUsage = async (req, adminToken) => {
+  const [users] = await db.user.find(
+    [
+      sql`users.data->>'stripeProductId' IN ('growth_1', 'scale_1', 'prod_O9XtHhI6rbTT1B','prod_O9XtcfOSMjSD5L')`,
+    ],
+    {
+      limit: 9999999999,
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts:26]: This limit will be forced to 1000 internally. We should ideally make this paginated or otherwise create a way to force a query without a limit (but make sure you don't allow _users_ to make that)

File: packages/api/src/controllers/stripe.ts [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+      useReplica: true,
+    }
+  );
+
+  const [oldProPlanUsers] = await db.user.find(
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts:34]: Can you add a `// TODO` somewhere to remove this temp/test plan logic?

File: packages/api/src/controllers/stripe.ts [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+  const payAsYouGoUsers = [...users, ...oldProPlanUsers];
+
+  let updatedUsers = [];
+  for (const user of payAsYouGoUsers) {
+    const userSubscription = await req.stripe.subscriptions.retrieve(
+      user.stripeCustomerSubscriptionId
+    );
+
+    const billingCycleStart = userSubscription.current_period_start * 1000; // 1685311200000 // Test date
+    const billingCycleEnd = userSubscription.current_period_end * 1000; // 1687989600000 // Test date
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: can remove comments

File: packages/api/src/controllers/stripe.ts [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+      {} as Record<string, string>
+    );
+
+    try {
+      // Invoice items based on overusage
+      await Promise.all(
+        products[user.stripeProductId].usage.map(async (product) => {
+          if (product.name === "Transcoding") {
+            await req.stripe.subscriptionItems.createUsageRecord(
+              subscriptionItemsByLookupKey["transcoding_usage"],
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Ideally we would use the lookup keys in the config instead of hardcoded strings. Probably fine for now.

File: packages/api/src/controllers/stripe.ts [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+    // create a map of subscription items by their lookup keys
+    const subscriptionItemsByLookupKey = subscriptionItems.data.reduce(
+      (acc, item) => {
+        acc[item.price.lookup_key] = item.id;
+        return acc;
+      },
+      {} as Record<string, string>
+    );
+
+    try {
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts:48]: Would be nice to wrap the whole iteration with the `try/catch` to make sure we don't fail-fast on errors in the above calls. Ideally could also move the `try` body to a helper func so we can have a clearer `try/call/catch/updateResult` logic here

File: packages/api/src/controllers/stripe.ts [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+  return {
+    updatedUsers: updatedUsers,
+  };
+};
+
+const calculateOverUsage = async (product, usage) => {
+  const limits = {
+    transcoding: product?.usage[0].limit,
+    streaming: product?.usage[1].limit,
+    storage: product?.usage[2].limit,
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Better rely on the config `name` instead of fixed indexes

File: packages/api/src/controllers/stripe.ts [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+  };
+
+  const overUsage = {
+    TotalUsageMins: Math.max(usage?.TotalUsageMins - limits.transcoding, 0),
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: **nit:** Might make sense to remove all the `?` in this func since the objs are never null

File: packages/api/src/controllers/stripe.ts [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+      headers: {
+        Authorization: `Bearer ${adminToken}`,
+      },
+    }
+  ).then((res) => res.json());
+
+  return usage;
+};
+
+// Webhook handler for asynchronous events.
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Can you update this comment to explain when/by whom this is called?

File: packages/api/src/controllers/stripe.ts [hunk 12]
```diff
@@ -0,0 +1,720 @@
...
+    // Notify via email
+    const invoice = event.data.object;
+    const [users] = await db.user.find(
+      { stripeCustomerId: invoice.customer },
+      { useReplica: false }
+    );
+
+    // notify help@livepeer.org
+    await sendgridEmailPaymentFailed({
+      email: "help@livepeer.org",
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts:271]: Might be worth a CLI arg for this.

Also, I think the implementation isn't this an array?
```ts
  const [supportName, supportEmail] = email;
```

File: packages/api/src/controllers/stripe.ts [hunk 13]
```diff
@@ -0,0 +1,720 @@
...
+      userId: users[0].id,
+      invoiceId: invoice.id,
+    });
+  }
+
+  // Return a response to acknowledge receipt of the event
+  return res.sendStatus(200);
+});
+
+async function sleep(millis) {
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts:282]: We already have this on an `utils` file, can reuse that

File: packages/api/src/controllers/stripe.ts [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+
+      // sleep for a 200 ms to get around stripe rate limits
+      await sleep(200);
+    }
+  }
+
+  res.json(users);
+});
+
+app.post("/migrate-personal-users", async (req, res) => {
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: This should be admin-only (`authorizer({anyAdmin: true})` or sth like that)

Edit: maybe its fine cause theres a requried secret below

File: packages/api/src/controllers/stripe.ts [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+
+    if (currentUsers.length === 0) {
+      keepGoing = false;
+      continue;
+    }
+
+    users = users.concat(currentUsers);
+
+    for (let index = 0; index < currentUsers.length; index++) {
+      let user = currentUsers[index];
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: **nit:** can be for-of
```suggestion
    for (const user of currentUsers) {
```

File: packages/api/src/controllers/stripe.ts [hunk 16]
```diff
@@ -0,0 +1,720 @@
...
+
+  const batchSize = 100;
+  let cursor = null;
+  let users = [];
+  let keepGoing = true;
+  let firstUserId = null;
+
+  let migratedUsers = [];
+
+  while (keepGoing) {
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Can consider removing this outer loop and calling the API multiple times instead. Can make the `batchSize` configurable via query-string to fine-tune it during execution

File: packages/api/src/controllers/stripe.ts [hunk 17]
```diff
@@ -0,0 +1,720 @@
...
+  let cursor = null;
+  let users = [];
+  let keepGoing = true;
+
+  let migratedUsers = [];
+  let firstUserId = null;
+
+  while (keepGoing) {
+    const [currentUsers, newCursor] = await db.user.find(
+      [sql`users.data->>'stripeProductId' = 'prod_1'`],
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Might make sense to check the `suspended` field of users as well (soccer)

File: packages/api/src/controllers/stripe.ts [hunk 18]
```diff
@@ -0,0 +1,720 @@
...
+          stripeCustomerPaymentMethodId: null,
+        });
+
+        // sleep for a 200 ms to get around stripe rate limits
+        await sleep(200);
+
+        // Update user's newStripeProductId to null
+        await db.user.update(user.id, {
+          newStripeProductId: null,
+        });
```
Comment (victorges) [inline packages/api/src/controllers/stripe.ts]: Move to the same update above

File: packages/api/src/controllers/usage.ts [hunk 19]
```diff
@@ -32,6 +34,19 @@ app.post(
   async (req, res) => {
     let { fromTime, toTime } = req.query;
 
+    const authHeader = req.headers.authorization;
+    let token = "";
+    if (authHeader) {
+      token = authHeader.split(" ")[1];
+    }
```
Comment (victorges) [inline packages/api/src/controllers/usage.ts]: You can probably just use req.token (and even avoid the temproary var, just sending as arg below)
```suggestion
    const token = req.token;
```

File: packages/api/src/controllers/usage.ts [hunk 20]
```diff
@@ -32,6 +34,19 @@ app.post(
...
+
+    // New automated billing usage report
+    let result = await reportUsage(req, token);
+
+    res.status(200);
+    res.json(result);
+    return;
```
Comment (victorges) [inline packages/api/src/controllers/usage.ts]: Delete the code below instead

File: packages/api/src/controllers/user.ts [hunk 21]
```diff
@@ -908,7 +894,7 @@ app.post(
     );
 
     // Temporarily skip updating the subscription if user is selecting a new plan
-    if (
+    /*if (
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:909]: Remove commented code

File: packages/api/src/controllers/user.ts [hunk 22]
```diff
@@ -963,25 +949,63 @@ app.post(
 
     */
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:962]: Remove commented code above as well

File: packages/api/src/controllers/user.ts [hunk 23]
```diff
@@ -963,25 +949,63 @@ app.post(
...
-            deleted: true,
-            clear_usage: true,
-            price: item.price.id,
-          })),
-          ...items.data.map((item) => ({
-            price: item.id,
-          })),
-        ],
+    let updatedSubscription;
+    if (products[payload.stripeProductId].deprecated) {
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:965]: We might not need this if our frontend is the only one calling

File: packages/api/src/controllers/user.ts [hunk 24]
```diff
@@ -963,25 +949,63 @@ app.post(
...
+              clear_usage: true,
+              price: item.price.id,
+            })),
+            ...items.data.map((item) => ({
+              price: item.id,
+            })),
+          ],
+        }
+      );
+    } else {
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:985]: TBC here

File: packages/api/src/controllers/stripe.ts (hunk 17 above)
Comment (victorges) [reply packages/api/src/controllers/stripe.ts]: Tho keep in mind suspended users may be unsuspended. Maybe we want to migrate them anyway

File: packages/api/src/controllers/user.ts [hunk 25]
```diff
@@ -963,25 +949,63 @@ app.post(
...
+      let payAsYouGoItems = [];
+      if (products[payload.stripeProductId].payAsYouGo) {
```
Comment (victorges) [inline packages/api/src/controllers/user.ts]: might wanna be safer with a `?` on the product, or first check if the product exists before using and returning an explicit error if not.

File: packages/api/src/controllers/user.ts [hunk 26]
```diff
@@ -963,25 +949,63 @@ app.post(
...
+        payload.stripeCustomerSubscriptionId,
+        {
+          billing_cycle_anchor: "now", // reset billing anchor when updating subscription
+          items: [
+            ...subscriptionItems.data.map((item) => {
+              // Check if the item is metered
+              const isMetered = item.price.recurring.usage_type === "metered";
+              return {
+                id: item.id,
+                deleted: true,
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:1008]: What does this mean?

File: packages/api/src/controllers/user.ts [hunk 27]
```diff
@@ -1026,6 +1050,31 @@ app.post(
   }
 );
 
+app.post(
+  "/retrieve-upcoming-invoice",
```
Comment (victorges) [inline packages/api/src/controllers/user.ts:1066]: Should this be a `GET /user/invoice/:stripeCustomerId` instead? 
Or maybe even just `GET /user/invoice` and the API just uses`req.user.stripeCustomerId` to grab the invoice.

File: packages/www/components/PlanForm/index.tsx [hunk 28]
```diff
@@ -428,12 +428,12 @@ const PlanForm = ({
                 <Box css={{ mt: "$4" }}>
                   <Text variant="neutral">
                     You are currently using the{" "}
-                    {products[user.newStripeProductId]?.name ||
-                      products[stripeProductId]?.name}{" "}
+                    {products[stripeProductId]?.name ||
+                      products[user.newStripeProductId]?.name}{" "}
```
Comment (victorges) [inline packages/www/components/PlanForm/index.tsx:438]: Is there any case where `products[stripeProductId]?.name` won't exist? Maybe the old order still makes sense as users get migrated to the new plans and have the `newStripeProductId` cleared out (then we can remove this logic completely).

File: packages/www/components/Plans/index.tsx [hunk 29]
```diff
@@ -251,19 +251,19 @@ const Plans = ({
...
               <Heading as="h3" size="3" css={{ mb: "$3", fontWeight: 600 }}>
-                {products["hacker_1"].name}
+                {products["prod_O9XuIjn7EqYRVW"].name}
               </Heading>
               <Box css={{ mb: "$4", fontSize: "$2" }}>Free</Box>
               <PlanForm
                 text={
                   dashboard
-                    ? newStripeProductId === "hacker_1"
+                    ? stripeProductId === "hacker_1"
```
Comment (victorges) [inline packages/www/components/Plans/index.tsx]: Is this ever true? Didn't you say that the actual production plans will be in the `prod_xxx` format? Same in all similar comparisons below.

File: packages/www/components/UpcomingInvoiceTable/index.tsx [hunk 30]
```diff
@@ -1,70 +1,141 @@
...
-          <Td>${transcodingPrice} / min</Td>
-          <Td>
-            {usage &&
-              `$${(
-                (usage.sourceSegmentsDuration / 60) *
-                transcodingPrice
-              ).toFixed(2)}`}
-          </Td>
-        </Tr>
+    <>
```
Comment (victorges) [inline packages/www/components/UpcomingInvoiceTable/index.tsx]: I think this surrounding `<> </>` is unnecessary since you only have 1 element inside. It's only necessary to create a fake parent when you want to return a list of elements from a react component. But I'm not an expert here.

File: packages/www/components/UsageSummary/index.tsx [hunk 31]
```diff
@@ -84,13 +101,28 @@ const UsageSummary = () => {
...
+        usage = {
+          TotalUsageMins: 0,
+          DeliveryUsageMins: 0,
+          StorageUsageMins: 0,
+        },
+      ] = await getBillingUsage(fromTime, toTime);
+
       if (res.status == 200) {
         setUsage(usage);
+        doCaculateOverUsage(usage);
```
Comment (victorges) [inline packages/www/components/UsageSummary/index.tsx]: Did you miss an `await`?

File: packages/www/components/UsageSummary/index.tsx [hunk 32]
```diff
@@ -101,10 +133,73 @@ const UsageSummary = () => {
...
+        TotalUsageMins: Math.max(usage?.TotalUsageMins - limits.transcoding, 0),
+        DeliveryUsageMins: Math.max(
+          usage?.DeliveryUsageMins - limits.streaming,
+          0
+        ),
+        StorageUsageMins: Math.max(usage?.StorageUsageMins - limits.storage, 0),
+      };
+
+      return overUsage;
+    };
```
Comment (victorges) [inline packages/www/components/UsageSummary/index.tsx]: Don't we have this same logic on the API? Feels like it should be exposed and called from here, not re-calculated in the frontend. 

An alternative could be to just fetch (directly or indirectly) the stripe state here? We should avoid making too many calls to BigQuery which is expensive, so it would be good if we could just re-use the state from the cronjob instead of re-querying and calculating on the dashboard on every reload.

You might want to consider leaving this as a future optimization as well, can you make sure to keep this tracked somewhere? For example can be tasks on the Linear projects for all these bits we talked about.

Maybe the same for the function below, though not sure if I saw the same thing on the backend.

File: packages/www/hooks/use-api/endpoints/user.ts [hunk 33]
```diff
@@ -227,9 +235,34 @@ export const getBillingUsage = async (
...
+      } as BillingUsageData | ApiError,
+    ];
+  }
+
   return [res, usage as BillingUsageData | ApiError];
 };
 
+export const getUpcomingInvoice = async (
+  stripeCustomerId: string
+): Promise<[Response, any | ApiError]> => {
```
Comment (victorges) [inline packages/www/hooks/use-api/endpoints/user.ts:250]: Can you add proper typing (no `any` for our own stuff) if/when you have a chance? Seems like there's already an `UpcomingInvoice` type declared?

File: packages/www/pages/dashboard/billing/index.tsx [hunk 34]
```diff
@@ -63,12 +80,97 @@ const Billing = () => {
...
+      if (overusage) {
+        const oBill = await calculateOverUsageBill(overusage);
+        setOverUsageBill(oBill);
+        let [res, uInvoice] = await getUpcomingInvoice(user.stripeCustomerId);
+        setUpcomingInvoice(uInvoice?.invoices);
+        setUpcomingInvoiceTotal((uInvoice?.invoices.total / 100) | 0);
       }
     };
+
+    const calculateOverUsage = async (usage) => {
```
Comment (victorges) [inline packages/www/pages/dashboard/billing/index.tsx:112]: Is this the same code from the other component? If it's not easy to re-use the backend logic, can you move it to a helper and avoid repeating it between the 2 frontend components?

File: packages/www/pages/dashboard/usage/index.tsx [hunk 35]
```diff
@@ -47,7 +51,14 @@ const Usage = () => {
...
   const doSetTimeStep = async (ts: string) => {
     setTimestep(ts);
-    const [res, usage] = await getBillingUsage(from, to, null, ts);
+    const [
+      res,
+      usage = {
+        TotalUsageMins: 0,
+        DeliveryUsageMins: 0,
+        StorageUsageMins: 0,
+      },
```
Comment (victorges) [inline packages/www/pages/dashboard/usage/index.tsx:60]: I think this default is redundant with the one you already added to `getBillingUsage` impl?

File: packages/www/pages/dashboard/usage/index.tsx [hunk 36]
```diff
@@ -84,25 +95,37 @@ const Usage = () => {
...
+      toTime: any,
+      status: any
+    ) => {
+      fromTime = fromTime * 1000;
+      toTime = toTime * 1000;
 
-      doSetFrom(fromTime);
-      doSetTo(toTime);*/
+      // if subscription is cancelled, get current month data
+      if (status === "canceled") {
```
Comment (victorges) [inline packages/www/pages/dashboard/usage/index.tsx:107]: Hm, seems like this logic could be inside `getBillingUsage` instead?

File: packages/www/pages/dashboard/usage/index.tsx [hunk 37]
```diff
@@ -84,25 +95,37 @@ const Usage = () => {
...
+
+      let [res, usage] = await getBillingUsage(fromTime, toTime);
 
-      const [res, usage] = await getBillingUsage(
-        fromTime * 1000,
-        toTime * 1000
-      );
       if (res.status == 200) {
+        if (!usage) {
+          usage = {
```
Comment (victorges) [inline packages/www/pages/dashboard/usage/index.tsx:117]: Same for redundant default.

**Comment by victorges** [comment]
@gioelecerati Tested the frontend but I got 0 usage on all users I tried. Even though on the home page it even shows a billing amount that seems to imply some usage, when I click on "view billing" or "view usage details" inside that there is `0.0` in all items. Not sure what is going on there.

This one is billing-test-7:
Screenshot 2023-07-11 at 20 14 24

But on home it showed `Upcoming invoice: $1775`

**Review by victorges — APPROVED** [review]
Didn't get to see it working E2E on the frontend as the usage was all 0, but it's probably some simple bug on the bigqueries or something.

LGTM either way, LFG!

File: packages/www/public/sitemap-0.xml
Comment (victorges) [inline packages/www/public/sitemap-0.xml:1]: FYI you can always just revert the changes in this file. It gets updated automatically when you build and i have no idea why, but no need to keep the changes afaik.

**Comment by gioelecerati** [comment]
> @gioelecerati Tested the frontend but I got 0 usage on all users I tried. Even though on the home page it even shows a billing amount that seems to imply some usage, when I click on "view billing" or "view usage details" inside that there is `0.0` in all items. Not sure what is going on there.
> 
> This one is billing-test-7: Screenshot 2023-07-11 at 20 14 24
> 
> But on home it showed `Upcoming invoice: $1775`

It's normal you are getting 0 as usage becase we removed the fixed timestamps with the june billing cycle (which is where the test usage data is residing). So July is empty!

File: packages/api/src/controllers/user.ts (hunk 26 above)
Comment (gioelecerati) [reply packages/api/src/controllers/user.ts:1008]: If the plan is metered (e.g. growth or scale), we clear the usage on plan update, because different plans has different overusage limits

File: packages/www/components/Plans/index.tsx (hunk 29 above)
Comment (gioelecerati) [reply packages/www/components/Plans/index.tsx]: Yes, they are staging product, just updated all the IDs to prod product for the deploy

File: packages/api/src/controllers/user.ts (hunk 27 above)
Comment (gioelecerati) [reply packages/api/src/controllers/user.ts:1066]: This api is not getting used any way for now, so I will leave it like this and update it later
